lane.py

```diff
@@ -1393,34 +1398,256 @@ def search_target(self):
+            qu = qu.add_columns(lane_id_field)
+        return qu
+
+    def _restrict_clause_to_window(self, clause, work_model, target_size):
+        """Restrict the given SQLAlchemy clause so that it matches
+        approximately `target_size` items.
+        """
+        if clause is None:
+            return clause
+        window_start, window_end = self.featured_window(target_size)
+        if window_start > 0 and window_start < 1:
+            clause = and_(
+                clause, 
+                work_model.random <= window_end,
+                work_model.random >= window_start
+            )
+        return clause
+
+    def _fill_parent_lane(self, additional_needed, unused_by_tier,
+                          used_by_tier, previously_used):
+        """Yield up to `additional_needed` randomly selected items from
+        `unused_by_tier`, falling back to `used_by_tier` if necessary.
+
+        :param unused_by_tier: A dictionary mapping quality tiers to
+        lists of unused MaterializedWork items. Because the same book
+        may have shown up as multiple MaterializedWork items, it may
+        show up as 'unused' here even if another occurance of it has
+        been used.
+
+        :param used_by_tier: A dictionary mapping quality tiers to lists
+        of previously used MaterializedWork items. These will only
+        be chosen once every item in unused_by_tier has been chosen.
+
+        :param previously_used: A set of work IDs corresponding to
+        previously selected MaterializedWork items. A work in
+        `unused_by_tier` will be treated as actually having been used
+        if its ID is in this set.
         """
-        # This takes care of all of the children.
-        works_and_lanes = super(Lane, self).groups(_db)
-
-        if not works_and_lanes:
-            # The children of this Lane did not contribute any works
-            # to the groups feed. This means there should not be
-            # a groups feed in the first place -- we should send a list
-            # feed instead.
-            return works_and_lanes
-
-        # The children of this Lane contributed works to the groups
-        # feed, which means we need an additional group in the feed
-        # representing everything in the Lane (since the child lanes
-        # are almost never exhaustive).
-        lane = _db.merge(self)
-        works = lane.featured_works(_db)
-        for work in works:
-            works_and_lanes.append((work, lane))
-        return works_and_lanes
-           
+        if not additional_needed:
+            return
+        additional_found = 0
+        for by_tier in unused_by_tier, used_by_tier:
+            # Go through each tier in decreasing quality order.
+            for tier in sorted(by_tier.keys(), key=lambda x: -x):
+                mws = by_tier[tier]
+                random.shuffle(mws)
+                for mw in mws:
+                    if (by_tier is unused_by_tier 
```

Comment:
should this be `in`?

Reply:
No, `is` is correct. We're looking in both `unused_by_tier` and `used_by_tier`, and this code only runs when we're looking in `unused_by_tier`.